src/frontend/serializable_models/state.ts

```diff
@@ -0,0 +1,154 @@
+    costFn?: calder.CostFn;
+    vectorField?: Float32Array;
+    guidingCurves?: List<calder.GuidingCurveInfo>;
+    selectedCurve?: number | null;
+    generating?: boolean;
+    pencilLine?: { x: number; y: number }[];
+};
+
+export class State implements Serializable<BakedState> {
+    public generator?: calder.Generator;
+    public source?: string;
+    public model?: calder.Model;
+    public costFnParams?: List<calder.GuidingCurve>;
+    public costFn?: calder.CostFn;
+    public vectorField?: Float32Array;
+    public guidingCurves?: List<calder.GuidingCurveInfo>;
+    public selectedCurve?: number | null;
+    public generating?: boolean;
+    public pencilLine?: { x: number; y: number }[];
+
+    /**
+     * constructor creates a new blank Serializable State object.
+     */
+    constructor() {
+        this.clearState()
+    }
+
+    /**
+     * serialize serializes the current serializedObject into a JSON compliant
+     * string to be stored in localstorage.
+     */
+    serialize(): string {
+        const serializedObject: SerializableState = {};
+        serializedObject.source = this.source;
+
+        serializedObject.costFnParams = this.costFnParams &&
+            this.costFnParams.toJS().map((curve) => {
+                return {
+                    distanceMultiplier: curve.distanceMultiplier,
+                    alignmentMultiplier: curve.alignmentMultiplier,
+                    alignmentOffset: curve.alignmentOffset,
+                    bezier: curve.bezier.points
+                };
+            });
+
+        return JSON.stringify(serializedObject);
+    }
+
+    /**
+     * deserialize takes a serialized representation of the object State as a
+     * JSON string and updates the properties of the object with the values
+     * represented in the string.
+     */
+    deserialize(serialized: string) {
+        // Deserialize the stored JSON object.
+        const serializedObject = <SerializableState>JSON.parse(serialized);
+
+        // Clear the undo and redo stacks for the editor.
+        clearUndoRedo();
+
+        // Clear the current properties of the State object.
+        this.clearState()
+
+        this.source = serializedObject.source;
+        this.costFnParams = serializedObject.costFnParams &&
+            List(serializedObject.costFnParams.map((curve: SerializableGuidingCurve) => {
+                return {
+                    distanceMultiplier: curve.distanceMultiplier,
+                    alignmentMultiplier: curve.alignmentMultiplier,
+                    alignmentOffset: curve.alignmentOffset,
+                    bezier: new Bezier(curve.bezier)
+                };
+            }));
+
+        // TODO: document what this does.
+        freshStateCallbacks.forEach((callback) => callback());
+    }
+
+    /**
+     * asBakedType returns a representation of the State object as a TypeScript
+     * type.
+     */
+    asBakedType(): BakedState {
+        return {
```

Comment:
Out of curiosity, does `return this` work here? It might mess up the `{ ...state.asBakedType() }` later if it also tries to copy the methods and not just the properties, I'm not sure if the spread operator does that or not

Reply:
Ahh, I'm not sure either. I'll look into it 👍